flow/layers/physical_shape_layer.cc

```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
-    FML_DCHECK(context.top_entity());
-    FML_DCHECK(retained_node.session() == context.session());
-    context.top_entity()->entity_node().AddChild(retained_node);
-    return;
-  }
+  // If there is embedded Fuchsia content in the scene (a ChildSceneLayer),
+  // PhysicalShapeLayers that appear above the embedded content will be turned
+  // into their own Scenic layers.
+  if (child_layer_exists_below_) {
+    // Reset paint bounds, because shadows are not rendered in this case
+    // (see comment in Paint()).
+    set_paint_bounds(path_.getBounds());
+
+    float global_scenic_elevation = context.GetGlobalElevationForNextScenicLayer();
+    float local_scenic_elevation_ =
+        global_scenic_elevation - context.scenic_elevation();
+    float z_translation = -local_scenic_elevation_;
+
+    // Retained rendering: speedup by reusing a retained entity node if
+    // possible. When an entity node is reused, no paint layer is added to the
+    // frame so we won't call PhysicalShapeLayer::Paint.
+    LayerRasterCacheKey key(unique_id(), context.Matrix());
+    if (context.HasRetainedNode(key)) {
+      TRACE_EVENT_INSTANT0("flutter", "retained layer cache hit");
+      scenic::EntityNode* retained_node = context.GetRetainedNode(key);
+      FML_DCHECK(context.top_entity());
+      FML_DCHECK(retained_node->session() == context.session());
 
-  TRACE_EVENT_INSTANT0("flutter", "cache miss, creating");
-  // If we can't find an existing retained surface, create one.
-  SceneUpdateContext::Frame frame(context, frameRRect_, color_, SK_AlphaOPAQUE,
-                                  elevation_, total_elevation_, this);
-  for (auto& layer : layers()) {
-    if (layer->needs_painting()) {
-      frame.AddPaintLayer(layer.get());
+      // Re-adjust the elevation.
+      retained_node->SetTranslation(0.f, 0.f, z_translation);
+
+      context.top_entity()->entity_node().AddChild(*retained_node);
+      return;
+    }
+
+    TRACE_EVENT_INSTANT0("flutter", "cache miss, creating");
+    // If we can't find an existing retained surface, create one.
+    SceneUpdateContext::Frame frame(context, frameRRect_, SK_ColorTRANSPARENT,
+                                    sk_float_round2int(context.alphaf() * 255),
+                                    "flutter::PhysicalShapeLayer",
+                                    z_translation, this);
+
+    frame.AddPaintLayer(this);
+
+    // Node: UpdateSceneChildren needs to be called here so that |frame| is still
+    // in scope (and therefore alive) while UpdateSceneChildren is being called.
+    if (children_need_system_compositing_) {
+      UpdateSceneChildren(context);
+    }
+  } else {
+    if (children_need_system_compositing_) {
+      UpdateSceneChildren(context);
     }
   }
+}
+
+void PhysicalShapeLayer::UpdateSceneChildren(SceneUpdateContext& context) {
+  float scenic_elevation = context.scenic_elevation();
+  context.set_scenic_elevation(scenic_elevation + local_scenic_elevation_);
```

Comment:
If you do this line in PhysicalShapeLayer::UpdateScene instead, you won't need to store `local_scenic_elevation_` as a member variable.

Reply:
i moved this because UpdateSceneChildren is called in two different places and this removed some duplicated code. But I think I can put the elevation logic at the very top and very bottom of UpdateScene which would eliminate the duplication and need for a variable, at the cost of a slight reduction in understandability.

Reply:
Removed local_scenic_elevation_.